data-model/src/main/java/org/hypertrace/core/datamodel/shared/ApiNode.java

```diff
@@ -3,46 +3,46 @@
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
-import org.hypertrace.core.datamodel.Event;
-import org.hypertrace.core.datamodel.StructuredTrace;
+import org.apache.avro.generic.GenericRecord;
 
-public class ApiNode {
+public class ApiNode<T extends GenericRecord> {
 
-  private final StructuredTrace trace;
-  private final Event entryApiBoundaryEvent;
-  private final List<Event> exitApiBoundaryEvents;
+  private final T entryApiBoundaryEvent;
+  private final List<T> exitApiBoundaryEvents;
+  private final T headSpan;
+  private final List<T> apiNodeEvents;
 
-  public ApiNode(StructuredTrace trace, Event apiEntryEvent, List<Event> exitEvents) {
-    this.trace = trace;
+  public ApiNode(T headSpan, List<T> apiNodeEvents, T apiEntryEvent, List<T> exitEvents) {
```

Comment:
What are apiNodeEvents? Probably, just events, I am assuming. Do we want to just call them as events?

Reply:
nit: Since everything is suffixed as event, we can probably rename `headSpan` to `rootEvent`?

Reply:
Can we also have a JavaDoc for the constructor? It would become easier to understand what all arguments need to be passed in the constructor

Reply:
yes. I'd vote to rename `apiEvents` to `events`

Reply:
> nit: Since everything is suffixed as event, we can probably rename `headSpan` to `rootEvent`?

I will change it to headEvent. I think `rootEvent` should be reserved for the root event in a full trace.

Reply:
I am confused. What's the difference between `headEvent` and `rootEvent`?

Reply:
Just terminology. If you see `rootEvent` you think about the first span in a full trace. Api Trace is our own concept and if this is named rootEvent, then people will think it corresponds to the root event in the full trace. That's why I'm not using it.

Reply:
But, weren't we getting the first span in the full trace

`return trace.getEventList().get(0);`

Reply:
event list was a subset of the full trace. trace was not the full trace but the api trace with just the events for this api trace.